File: CANOpen Protocol/src/canopen_serialize.py
import struct

def serialize(node_id, command, index, subindex, data):
    # Convert the data to bytes
    data_bytes = data.encode('utf-8')
    data_length = len(data_bytes)
    
    # Create the CANopen message
    # struct.pack creates a packed binary string from the given values
    # 'B' stands for unsigned char (1 byte)
    # 'H' stands for unsigned short (2 bytes)
    message = struct.pack('B B H B B', node_id, command, index, subindex, data_length) + data_bytes
    
    return message

# Example data to be serialized
node_id = 10
command = 0xA1  # Write command
index = 0x2000  # Index in CANopen dictionary
subindex = 1    # Subindex in CANopen dictionary
data = "Adiyat Rahman"  # Data to be sent

# Serialize the data
message = serialize(node_id, command, index, subindex, data)
print("Serialized CANopen message in Python:")
# Print the serialized message as a hexadecimal string
print(message.hex())
hex_string = str(message.hex())
print(hex_string)

# convert the hex string into an array of bytes
byte_array = [int(hex_string[i:i+2], 16) for i in range(0, len(hex_string), 2)]
c_byte_array = "{" + ", ".join([f"0x{byte:02X}" for byte in byte_array]) + "}"
print(c_byte_array)

File: CANOpen Protocol/src/canopen_serialize.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Function to serialize data into a CANopen-like message
void serialize(unsigned char node_id, unsigned char command, unsigned short index, unsigned char subindex, const char* data) {
    unsigned char data_length = strlen(data);  // Get the length of the data
    unsigned char message[6 + data_length];    // Create a message array of appropriate size

    // Pack the fixed part of the message
    message[0] = node_id;                       // Node ID (1 byte)
    message[1] = command;                       // Command (1 byte)
    message[2] = index & 0xFF;                  // Lower byte of index (2 bytes)
    message[3] = (index >> 8) & 0xFF;           // Higher byte of index (2 bytes)
    message[4] = subindex;                      // Subindex (1 byte)
    message[5] = data_length;                   // Data length (1 byte)

    // Copy the variable length data part into the message
    memcpy(&message[6], data, data_length);     // Copy data into the message array

    // Print the serialized message in hex format
    printf("Serialized CANopen message in C:\n");
    for (int i = 0; i < 6 + data_length; i++) {
        printf("%02x", message[i]);              // Print each byte as a two-digit hex value
    }
    printf("\n");
}

int main() {
    unsigned char node_id = 1;                  // Node ID
    unsigned char command = 0x01;               // Write command
    unsigned short index = 0x2000;              // Index in CANopen dictionary
    unsigned char subindex = 1;                 // Subindex in CANopen dictionary
    const char* data = "Adiyat Rahman";         // Data to be sent

    // Serialize the data
    serialize(node_id, command, index, subindex, data);

    return 0;
}
